### PR TITLE
Compare target pwm value to pwm1_min and pwm1_max instead of fan rpm

### PR DESCRIPTION
**Pull request by r15ch13**
See https://github.com/Eraden/amdgpud/issues/4#issuecomment-882628078

### CODE REVIEW

**Comment by BoostCookie** [comment]
Why did you choose 1500 as a default for fan1_max?

**Comment by r15ch13** [comment]
Hm, for safety I guess. 😄

**Comment by BoostCookie** [comment]
With these changes the speed values in the config file need to be between 4 and 255 right? I think it would be nicer to have the speed values in the config as a percentage (I think that's how it was initially intended) and the program converts it to the corresponding pwm value. But lets wait and see what the maintainer says.

**Comment by Eraden** [comment]
4 is just lowest safest value which works. This is native driver bug not actual value. It's just u8 with min 0 and max 255 but for some reason below 4 does nothing.

**Comment by Eraden** [comment]
> Hm, for safety I guess. 😄
> It could also be 0, because it will return that value if `fan1_max` couldn't be read.
> But setting the max value too low doesn't seem to be a good idea if something goes wrong.

This is good for now and I'm really appreciate your contribution.

I need to ask to add TODO with comment to add array of max speed per card or any other solution. Max speed for cards will change in time and part of them is water cooled.

If this would be percentage or other vendor min-max fixed value this would be much easier.

Note: I'm on holiday far from PC for next 10 dats. Please wait for approval 🙏

**Comment by r15ch13** [comment]
It should work by changing the `speed_for_temp()` function to return the PWM calculated from the config percentage.
```
pwm = (config speed value / 100.0 * 255.0) as int

75% / 100 * 255 = 191 PWM
50% / 100 * 255 = 127 PWM
25% / 100 * 255 = 63 PWM
```

File: src/main.rs [hunk 1]
```diff
@@ -288,13 +323,13 @@ fn service(config: Config) -> std::io::Result<()> {
         for controller in controllers.iter_mut() {
             let gpu_temp = controller.hw_mon.gpu_temp().unwrap_or_default();
 
-            let speed = config.speed_for_temp(gpu_temp);
-            if controller.hw_mon.fan_min() > speed || controller.hw_mon.fan_max() < speed {
+            let target_pwm = config.speed_for_temp(gpu_temp);
```
Comment (r15ch13) [inline src/main.rs:326]: This is enough to convert the percentage to PWM values:
```suggestion
            let speed = config.speed_for_temp(gpu_temp);
            let target_pwm = (speed as f32 * 2.55).round() as u32;
```

**Comment by BoostCookie** [comment]
> 4 is just lowest safest value which works. This is native driver bug not actual value. It's just u8 with min 0 and max 255 but for some reason below 4 does nothing.

Are you sure this is a bug? When I write something into `pwm1` my card actually responds, but the value I read out after that is not the same one I've written in. I don't know why that is, but still the whole range from 0 to 255 works. There's just some lowest value below which the fans don't spin anymore. Maybe that's where your 4 is?

This value is different for each card though and should not be part of the program. It's the user's task to find out where their fans stop spinning and write the config file accordingly.

> Writing percentage calculation may make more difficult tracking down possible bugs and may create redundant code

I think it would be really much nicer though. The actual granularity of controlling the fan is at 256. If we only allow whole percentages we are getting only 100 possible values out of the 256.

File: src/main.rs (hunk 1 above)
Comment (Eraden) [reply src/main.rs]: 1% would give 2,55 which is below 4 and would be ignored

**Comment by Eraden** [comment]
@BoostCookie please open for percentage issue. It's not I don't agree but this PR is fixing something other

File: src/main.rs [hunk 2]
```diff
@@ -82,6 +86,37 @@ impl HwMon {
...
+    pub fn pwm_min(&mut self) -> u32 {
+        if self.pwm_min.is_none() {
+            self.pwm_min = Some(
+                self.read("pwm1_min")
+                    .unwrap_or_default()
+                    .parse()
+                    .unwrap_or(0),
+            )
+        };
+        self.pwm_min.unwrap_or(0)
```
Comment (Eraden) [inline src/main.rs:98]: Please change to `unwrap_or_default()`

Comment (r15ch13) [reply src/main.rs:98]: This was copied from `fan_min()`. Shouldn't it be changed there too?

Comment (Eraden) [reply src/main.rs:98]: Yes, please change also there

File: src/main.rs [hunk 3]
```diff
@@ -289,12 +324,13 @@ fn service(config: Config) -> std::io::Result<()> {
             let gpu_temp = controller.hw_mon.gpu_temp().unwrap_or_default();
 
             let speed = config.speed_for_temp(gpu_temp);
-            if controller.hw_mon.fan_min() > speed || controller.hw_mon.fan_max() < speed {
+            let target_pwm = (speed as f32 * 2.55).round() as u32;
```
Comment (Eraden) [inline src/main.rs:327]: NOTE: @Eraden 

I didn't noticed this during review. Documentation need to be updated

**Pull request merged**